### PR TITLE
Integrate RAFT FlatIndex / Distances

### PR DESCRIPTION
**Pull request by cjnolet**
I've broken out the FlatIndex / Distances changes from #2521 into a separate PR to make things a litle easier to review. This does also include a couple other minor changes to the IVF Flat index which I used to make it easier to dispatch to the RAFT version. I can revert that change too if desired.

### CODE REVIEW

**Review by mdouze — COMMENTED** [review]
I defer to @wickedfoo and @alexanderguzhva for a detailed review. 
A high-level question: would it be possible to implement the `bfKnn` function in

https://github.com/facebookresearch/faiss/blob/main/faiss/gpu/GpuDistance.h

with RAFT?

**Comment by cjnolet** [comment]
> I defer to @wickedfoo and @alexanderguzhva for a detailed review.
> A high-level question: would it be possible to implement the bfKnn function in
> https://github.com/facebookresearch/faiss/blob/main/faiss/gpu/GpuDistance.h
> with RAFT?

Yes, absolutely, we can plug in there.

**Comment by cjnolet** [comment]
@mdouze looking a little more closely at the `GpuDistance.cu`, RAFT does have corresponding functions to compute this but it's not immediately clear to me how to enable the RAFT dispatch there since we've been using an option in the `GpuindexConfig` to allow the user to enable it at runtime. Another option could be to also include a `use_raft` member on the `GpuDistanceParams`.

File: faiss/gpu/GpuDistance.h [hunk 1]
```diff
@@ -124,6 +124,18 @@ struct GpuDistanceParams {
     /// Otherwise, an integer 0 <= device < numDevices indicates the device for
     /// execution
     int device;
+
+    /// Should the index dispatch down to RAFT?
+    bool use_raft = false;
+
+    bool should_use_raft() const {
```
Comment (mdouze) [inline faiss/gpu/GpuDistance.h:131]: why not access use_raft directly?

File: faiss/gpu/test/TestGpuIndexFlat.cpp [hunk 2]
```diff
@@ -73,6 +73,9 @@ void testFlat(const TestFlatOptions& opt) {
     faiss::gpu::GpuIndexFlatConfig config;
     config.device = device;
     config.useFloat16 = opt.useFloat16;
+#if defined USE_NVIDIA_RAFT
```
Comment (mdouze) [inline faiss/gpu/test/TestGpuIndexFlat.cpp:76]: shouldn't we test both with and without RAFT?

**Review by mdouze — COMMENTED** [review]
I am fine with having use_raft as a field that exists even if RAFT is not compiled in (in that case an error is raised when use_raft is set to true).

File: faiss/gpu/test/TestGpuIndexFlat.cpp (hunk 2 above)
Comment (cjnolet) [reply faiss/gpu/test/TestGpuIndexFlat.cpp:76]: Yes we should. We've just turned this on in the meantime so we can get all the tests passing from the RAFT side.

**Comment by mdouze** [comment]
FYI I started working on 
https://github.com/facebookresearch/faiss/pull/2722
the objective is to gain flexibility in installing whatever is needed on a vanilla ubuntu docker image to compile and test Faiss. 
I would like to work this out for the GPU contbuild so that RAFT passes the CI tests. 
We'll see later if RAFT GPU will go in its own tests or be merged with the regular GPU tests.

**Comment by cjnolet** [comment]
> We'll see later if RAFT GPU will go in its own tests

@mdouze just to clarify- do you mean separate test files do you mean using the preprocessor conditionals?

**Comment by mdouze** [comment]
@cjnolet I mean that we have one "Linux GPU" check currently in circle CI (that runs iff the CPU Linux tests are successful, which is why it does not appear in this diff). In addition we may add a "Linux GPU / RAFT", or include the RAFT version in the "Linux GPU" check. TBD

**Comment by mdouze** [comment]
Excellent! Next steps for me: integration in circleCI then move to fbsource (our internal source of truth).

**Comment by cjnolet** [comment]
@mdouze RAPIDS supports CUDA 11.2+. We are planning to release CUDA 12 support w/ the 23.04 release (in mid-April).

**Comment by algoriddle** [comment]
@cjnolet I'd like to help move this forward. I landed a cleanup of the contbuild script last week, in preparation for adding a RAFT-enabled flavour. First, I'd like to make sure that this PR doesn't break any of the non-RAFT builds. Would you mind rebasing on top of the current main, please? I just tried to do so, but got lost in the merge conflicts.

**Comment by algoriddle** [comment]
@cjnolet Thanks, it looks like the first blocker was that you needed a more recent cmake. I have just landed the required change for the build envs in main (commit a3d62129). Would you please rebase again?

**Comment by algoriddle** [comment]
@cjnolet I fixed two problems that triggered build or test failures, so that we can build without raft. However, when I try to build **with** raft, I get the following error:

```
$ ./build.sh
[...]
-- Auto detection of gpu-archs: 60
-- CPM: adding package raft@23.02.00 (bfknn)
[ 11%] Creating directories for 'raft-populate'
[ 22%] Performing download step (git clone) for 'raft-populate'
Cloning into 'raft-src'...
fatal: invalid reference: bfknn
CMake Error at raft-subbuild/raft-populate-prefix/tmp/raft-populate-gitclone.cmake:49 (message):
  Failed to checkout tag: 'bfknn'


make[2]: *** [CMakeFiles/raft-populate.dir/build.make:102: raft-populate-prefix/src/raft-populate-stamp/raft-populate-download] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/raft-populate.dir/all] Error 2
make: *** [Makefile:91: all] Error 2

CMake Error at /private/home/gsz/.conda/envs/faiss_from_source/share/cmake-3.23/Modules/FetchContent.cmake:1088 (message):
  Build step for raft failed: 2
Call Stack (most recent call first):
  /private/home/gsz/.conda/envs/faiss_from_source/share/cmake-3.23/Modules/FetchContent.cmake:1217:EVAL:2 (__FetchContent_directPopulate)
  /private/home/gsz/.conda/envs/faiss_from_source/share/cmake-3.23/Modules/FetchContent.cmake:1217 (cmake_language)
  build/cmake/CPM_0.35.6.cmake:963 (FetchContent_Populate)
  build/cmake/CPM_0.35.6.cmake:773 (cpm_fetch_package)
  build/cmake/CPM_0.35.6.cmake:299 (CPMAddPackage)
  build/_deps/rapids-cmake-src/rapids-cmake/cpm/find.cmake:167 (CPMFindPackage)
  cmake/thirdparty/get_raft.cmake:31 (rapids_cpm_find)
  cmake/thirdparty/get_raft.cmake:50 (find_and_configure_raft)
  CMakeLists.txt:69 (include)


-- Configuring incomplete, errors occurred!
See also "/private/home/gsz/github/facebookresearch/faiss/build/CMakeFiles/CMakeOutput.log".
```

**Comment by cjnolet** [comment]
@algoriddle I've made the change to this branch in my fork but it's not updating this PR, for some reason. I'm not sure I've ever seen this before. You can see two additional commits in the branch in my fork that are not in this PR: https://github.com/cjnolet/faiss/commits/integrate_raft_flat. Is this expected?

**Comment by algoriddle** [comment]
@cjnolet I think it's okay to use nightlies for now. I'm not sure why the PR isn't updating, it could be because we already imported it into fbsource (Meta internal) and that version slightly diverged, I'm not sure. We can apply the fix separately.

**Pull request closed without merging**